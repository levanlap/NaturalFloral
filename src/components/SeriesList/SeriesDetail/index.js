import React, { Component } from 'react'
import { StackNavigator } from 'react-navigation'
import PropTypes from 'prop-types'
import MatchAndStat from './scenes/MatchAndStat'
import RecentSeries from './scenes/RecentSeries'

export default class SeriesDetail extends Component {
  static propTypes = {
    navigation: PropTypes.object,
    getRoutes: PropTypes.func
  }

  _generateNavigator(item) {
    return StackNavigator({ // eslint-disable-line
      MatchAndStat: {
        screen: (props) => <MatchAndStat
          {...props} stackNavigation={this.props.navigation}
          item={item}
          getRoutes={this.props.getRoutes}
        />
      },
      RecentSeries: {
        screen: (props) => <RecentSeries {...props} />
      },
      RecentSeriesDetail: {
        screen: (props) => <MatchAndStat
          {...props}
          item={props.navigation.state.params.item}
          getRoutes={this.props.getRoutes}
        />
      }
    },
    {
      headerMode: 'none'
    })
  }

  render() {
    const Screen = this._generateNavigator(this.props.navigation.state.params.item)

    return <Screen />
  }
}
