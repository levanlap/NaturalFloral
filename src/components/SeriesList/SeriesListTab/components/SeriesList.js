import React, { PureComponent } from 'react'
import { StyleSheet, Image, FlatList, RefreshControl } from 'react-native'
import { Spinner, ListItem, Text, Body, View } from 'native-base'
import Tracker from '../../../../services/tracker'
import Store from '../../../../services/store'
import AdsBanner from '../../../AdsBanner'
import ScoreBoard from '../../../ScoreBoard'
import ListMatch from '../../../../components/ListMatch'

export default class ListSeries extends PureComponent {
  constructor() {
    super()
    this.state = {
      isFetching: true,
      page: 1,
      stickyHeaderIndices: [],
      seriesState: []
    }
  }

  componentWillMount() {
    if (this.props.initState) {
      this.setState(...this.props.initState)
    }
    this._onRefresh()
  }

  _onRefresh = () => {
    this.setState({ isFetching: true, page: 1 }, async () => {
      const { data } = await Store.getSeries(this.props.seriesType, this.state.page)
      const items = data
      const stickyHeaderIndices = []
      const loadImageJob = []
      const seriesState = []

      items.map((item, index) => {
        if (item.header) {
          stickyHeaderIndices.push(index)
        }
        if (item.left && item.right) {
          loadImageJob.push(Image.prefetch(item.left.logo))
          loadImageJob.push(Image.prefetch(item.right.logo))
        }
        if (item.series && (item.meta.id === 'history')) {
          this.callToGetSeriesDetail(item).then((data) => {
            let seriesDetail = data
            seriesState.push({
              series_id: item.series.series_id,
              show: false,
              seriesDetail
            })
          })
        }
      })
      Promise.all(loadImageJob)
      stickyHeaderIndices.push(0)
      this.setState({ isFetching: false, items, stickyHeaderIndices, seriesState })
    })
  }

  _onLoadMore = () => {
    if (this.state.isFetching || this.state.isLoadingMore || this.state.lastItem) return
    this.setState({ page: this.state.page + 1, isLoadingMore: true }, async () => {
      const { data, lastItem } = await Store.getSeries(this.props.seriesType, this.state.page)
      const items = this.state.items.concat(data)
      const stickyHeaderIndices = []
      const loadImageJob = []
      const seriesState = []

      items.map((item, index) => {
        if (item.header) {
          stickyHeaderIndices.push(index)
        }
        if (item.left && item.right) {
          loadImageJob.push(Image.prefetch(item.left.logo))
          loadImageJob.push(Image.prefetch(item.right.logo))
        }
        if (item.series && (item.meta.id === 'history')) {
          this.callToGetSeriesDetail(item).then((data) => {
            let seriesDetail = data
            seriesState.push({
              series_id: item.series.series_id,
              show: false,
              seriesDetail
            })
          })
        }
      })
      Promise.all(loadImageJob).catch(() => { })
      stickyHeaderIndices.push(0)
      this.setState({
        items, stickyHeaderIndices, isLoadingMore: false, lastItem, seriesState
      })
    })
  }

  async callToGetSeriesDetail(item) {
    const result = await Store.getSeriesDetail(item)
    return result
  }

  _viewDetail = (item) => {
    if (this.props.stackNavigation) {
      Tracker.fireEvent(
        `SeriesList-${this.props.navigation.state.routeName}`, 'SelectSeries',
        `${item.left.name} - ${item.right.name}`
      )
      this.props.stackNavigation.navigate('SeriesDetail', { item })
    }
  }

  _showMatch(item) {
    if (item.meta.id === 'history') {
      let seriesState = [...this.state.seriesState]
      let index = seriesState.findIndex((series) => series.series_id === item.series.series_id)
      if (seriesState[index]) {
        seriesState[index].show = !seriesState[index].show
      }
      this.setState({ seriesState })
    }
    else {
      this._viewDetail(item)
    }
  }

  _renderMatch(leftSide, item) {
    let seriesID = item.series.series_id
    if (this.state.seriesState.find((series) => series.series_id === seriesID && series.show)) {
      const current = this.state.seriesState.findIndex((series) => series.series_id === seriesID)
      const seriesDetail = this.state.seriesState[current].seriesDetail
      return (
        <View>
          <View style={styles.separator} />
          <ListMatch
            overview={item}
            stackNavigation={this.props.stackNavigation}
            seriesID={seriesID}
            listMatch={seriesDetail.games}
            teams={seriesDetail.teams}
            radiantId={leftSide.id}
          />
        </View>
      )
    }
  }
  _renderItem = ({ item }) => {
    const { left } = item
    if (item.header) {
      return (
        <ListItem itemDivider key={item.name} style={[styles.listViewItem, styles.header]}>
          <Text style={styles.headerText}>
            {item.name.toUpperCase()}
          </Text>
        </ListItem>
      )
    }
    else if (item.isAds) {
      return (
        <ListItem key={item.id} style={styles.listViewItem}>
          <AdsBanner />
        </ListItem>
      )
    }
    else if (item.isLoading) {
      return (
        <ListItem key={`loading-${item.id}`} style={[styles.listViewItem, { justifyContent: 'center' }]}>
          <Spinner color='#388E3C' />
        </ListItem>
      )
    }
    return (
      <ListItem key={item.id} style={[styles.listViewItem, styles.series]} onPress={() => {
        this._showMatch(item)
      }}>
        <Body>
          <ScoreBoard item={item} />
          {this._renderMatch(left, item)}
        </Body>
      </ListItem>
    )
  }
  render() {
    let items = this.state.items
    if (this.state.isLoadingMore) {
      items = items.concat([{ isLoading: true, id: Math.random() }])
    }
    return (
      <FlatList
        onEndReachedThreshold={0.8}
        onEndReached={this._onLoadMore}
        style={{ height: '100%' }}
        data={items}
        renderItem={this._renderItem}
        keyExtractor={(item) => item.id.toString()}
        stickyHeaderIndices={this.state.stickyHeaderIndices}
        refreshControl={
          <RefreshControl
            refreshing={this.state.isFetching}
            onRefresh={this._onRefresh}
            title='Pull to refresh'
            tintColor='#388E3C'
            titleColor='#959595'
          />
        }
      />
    )
  }
}

const styles = StyleSheet.create({
  separator: {
    backgroundColor: '#D7D7D7',
    height: 1,
    marginLeft: 8,
    marginRight: 8
  },
  listViewItem: {
    width: '100%',
    marginLeft: 0,
    paddingLeft: 0,
    paddingTop: 0,
    paddingBottom: 0,
    paddingRight: 0,
    marginRight: 0
  },
  header: {
    backgroundColor: '#E8E7E7',
    height: 24
  },
  headerText: {
    fontFamily: 'Roboto_medium',
    paddingLeft: 8,
    color: '#959595',
    fontSize: 11,
    fontWeight: 'bold'
  }
})
