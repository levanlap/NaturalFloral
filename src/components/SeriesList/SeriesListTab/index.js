import React from 'react'
import { View } from 'react-native'
import { createNavigator, createNavigationContainer, TabRouter } from 'react-navigation'
import {
  Container, Header,
  Left, Right, Body,
  Button, Text
} from 'native-base'
import { Svg } from 'expo'
import TabView from './components/TabView'
import SeriesList from './components/SeriesList'

const { Rect } = Svg

const Menu = () => (
  <Svg width={20} height={16} viewBox='0 0 20 16'>
    <Rect
      fill='#FFFFFF' width='20' height='4'
      rx='2' ry='2'
    />
    <Rect
      fill='#FFFFFF' y='6' width='15'
      height='4' rx='2' ry='2'
    />
    <Rect
      fill='#FFFFFF' y='12' width='10'
      height='4' rx='2' ry='2'
    />
  </Svg>
)

export default function SeriesListTab({ openMenu, navigation, game }) {
  const tabs = {}
  game.seriesTypes.map((seriesType) => {
    const extraProps = { game, seriesType, stackNavigation: navigation }
    tabs[seriesType.title] = {
      screen: (props) => (<SeriesList {...props} {...extraProps} />)
    }
  })

  const Tabs = createNavigationContainer(createNavigator(TabRouter(tabs))(TabView)) // eslint-disable-line

  return (
    <Container style={{ backgroundColor: '#E8E7E7' }}>
      <Header style={{ backgroundColor: '#388E3C' }}>
        <Left style={{ flex: 0.2 }}>
          <Button transparent onPress={openMenu}>
            <Menu />
          </Button>
        </Left>
        <Body>
          <Text style={{ fontFamily: 'Roboto_medium', color: '#FFFFFF' }}>{game.name}</Text>
        </Body>
        <Right style={{ flex: 0.2 }} />
      </Header>
      <View style={{ flex: 1 }}>
        <Tabs />
      </View>
    </Container>
  )
}
