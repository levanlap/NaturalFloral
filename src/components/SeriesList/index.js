import React, { Component } from 'react'
import { StackNavigator } from 'react-navigation'
import PropTypes from 'prop-types'
import SeriesListTab from './SeriesListTab'
import SeriesDetail from './SeriesDetail'

export default class SeriesListScreen extends Component {
  static propTypes = {
    navigation: PropTypes.object,
    game: PropTypes.object,
    getRoutes: PropTypes.func
  }

  _openMenu = () => {
    this.props.navigation.navigate('DrawerOpen')
  }

  _generateNavigator() {
    return StackNavigator({ // eslint-disable-line
      SeriesListTab: {
        screen: (props) => <SeriesListTab {...props} openMenu={this._openMenu} game={this.props.game} />
      },
      SeriesDetail: {
        screen: (props) => <SeriesDetail {...props} {...props} getRoutes={this.props.getRoutes} />
      }
    }, {
      headerMode: 'none'
    })
  }

  render() {
    const Screen = this._generateNavigator()

    return <Screen />
  }
}
