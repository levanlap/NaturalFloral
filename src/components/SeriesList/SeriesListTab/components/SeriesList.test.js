import React from 'react'
import SeriesList from './SeriesList'

import renderer from 'react-test-renderer'

it('renders without crashing', () => {
  const rendered = renderer.create(<SeriesList />).toJSON()
  expect(rendered).toBeTruthy()
})

it('render correctly', async () => {
  const tree = renderer.create(<SeriesList />).toJSON()
  expect(tree).toMatchSnapshot()
})
