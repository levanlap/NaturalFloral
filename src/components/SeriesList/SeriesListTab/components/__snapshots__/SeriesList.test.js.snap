// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<RCTScrollView
  data={undefined}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  invertStickyHeaders={undefined}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.8}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
      tintColor="#388E3C"
      title="Pull to refresh"
      titleColor="#959595"
    />
  }
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "height": "100%",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View />
</RCTScrollView>
`;
